### PR TITLE
[WIP] Use spans instead of single locations in AST

### PR DESCRIPTION
**Pull request by brendanzab**

### CODE REVIEW

File: base/src/ast.rs [hunk 1]
```diff
@@ -194,48 +194,48 @@ pub enum Pattern<Id: AstId> {
...
-        exprs: Vec<(Id::Untyped, Option<LExpr<Id>>)>,
+        exprs: Vec<(Id::Untyped, Option<SExpr<Id>>)>,
     },
     Lambda(Lambda<Id>),
-    Tuple(Vec<LExpr<Id>>),
-    Type(Vec<TypeBinding<Id::Untyped>>, Box<LExpr<Id>>),
-    Block(Vec<LExpr<Id>>),
+    Tuple(Vec<SExpr<Id>>),
+    Type(Vec<TypeBinding<Id::Untyped>>, Box<SExpr<Id>>),
+    Block(Vec<SExpr<Id>>),
```
Comment (Marwes) [inline base/src/ast.rs]: I think this is a bit to many constructors.

```rust
pub enum Expr<Id> {
    Atom(Id),
    Cons(SExpr<Id>, SExpr<Id>),
}
```
Should be enough no? :stuck_out_tongue:

File: base/src/pos.rs [hunk 2]
```diff
@@ -135,64 +135,62 @@ pub struct Span {
...
-#[derive(Copy, Clone, PartialEq, Debug)]
-pub struct Spanned<T> {
-    pub span: Span,
-    pub value: T,
-}
-
-impl<T: fmt::Display> fmt::Display for Spanned<T> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}: {}", self.span.start, self.value)
+    pub fn merge(self, other: Span) -> Option<Span> {
```
Comment (Marwes) [inline base/src/pos.rs]: Didn't see this used anywhere (yet)? Do you have a usecase in mind?

File: base/src/ast.rs (hunk 1 above)
Comment (brendanzab) [inline base/src/ast.rs]: Yes - I was wondering about that! `SpExpr`? :P

Comment (Marwes) [inline base/src/ast.rs]: Nah, `SExpr` is fine

File: base/src/pos.rs (hunk 2 above)
Comment (brendanzab) [inline base/src/pos.rs]: I was thinking that this would allow us to expand the spans during parsing, but we are only emitting located tokens right now....

**Comment by brendanzab** [comment]
Updated - I am not expanding the spans in the parser though, so I need to figure that out... :/

**Pull request closed without merging**